### ISSUE TITLE
[Phase 1] Foundation - Data Pipeline & Basic Detection

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#2]
## Summary

Create the data loading infrastructure and implement the first anomaly detector (Isolation Forest). This phase establishes the foundation for all subsequent work.

**Part of:** Meta Issue #1
**Dependencies:** None (starting phase)

## Objectives

1. Load and sample CSV datasets (KDD Cup 99 for testing)
2. Preprocess and normalize features
3. Implement Isolation Forest detector
4. Create project configuration structure

## Files to Create

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/__init__.py` | 1 | Create | Package marker |
| `src/ingestion/__init__.py` | 1 | Create | Package marker |
| `src/ingestion/data_loader.py` | 80 | Create | CSV loading with sampling |
| `src/ingestion/preprocessor.py` | 100 | Create | Normalization, missing values |
| `src/detection/__init__.py` | 1 | Create | Package marker |
| `src/detection/base_detector.py` | 60 | Create | Abstract base class |
| `src/detection/isolation_forest.py` | 80 | Create | Isolation Forest wrapper |
| `tests/test_data_loader.py` | 60 | Create | Data loader tests |
| `config/default.yaml` | 50 | Create | Configuration structure |
| `requirements.txt` | 15 | Create | Dependencies |

### Setup
- [ ] Create directory structure (`src/ingestion/`, `src/detection/`, `tests/`, `config/`, `data/`)
- [ ] Create `requirements.txt` with core dependencies
- [ ] Set up virtual environment and install deps

### Data Loading
- [ ] Implement `DataLoader.__init__()` with config parsing
- [ ] Implement `DataLoader.load_csv()` with validation
- [ ] Implement `DataLoader.load_chunks()` generator for large files
- [ ] Write unit tests for data loading

### Preprocessing
- [ ] Implement `Preprocessor.normalize()` using StandardScaler
- [ ] Implement `Preprocessor.handle_missing()` with median/mean/drop options
- [ ] Implement `Preprocessor.fit_transform()` pipeline method
- [ ] Write unit tests for preprocessing

### Detection Base
- [ ] Implement `BaseDetector` abstract class with `fit()`, `score_samples()`, `predict()`
- [ ] Add `is_fitted` state tracking
- [ ] Add docstrings with usage examples

### Isolation Forest
- [ ] Implement `IsolationForestDetector` wrapping sklearn
- [ ] Configure via YAML (`n_estimators`, `contamination`)
- [ ] Invert sklearn scores (sklearn: lower = anomaly; we want: higher = anomaly)
- [ ] Write unit tests verifying anomalies score higher

### Configuration
- [ ] Create `config/default.yaml` with data, preprocessing, detection sections
- [ ] Add YAML loading utility function

## Code Snippet: DataLoader

```python
# src/ingestion/data_loader.py
"""Data loading utilities for anomaly detection system."""

from pathlib import Path
from typing import Iterator, Optional
import pandas as pd


class DataLoader:
    """Load and sample datasets for anomaly detection."""

    def __init__(self, config: dict):
        self.config = config
        self.data_path = Path(config.get('data', {}).get('path', 'data/'))

    def load_csv(
        self,
        filename: str,
        sample_frac: Optional[float] = None
    ) -> pd.DataFrame:
        """
        Load CSV file with optional sampling.

        Args:
            filename: Name of CSV file in data directory
            sample_frac: Fraction of data to sample (0.0-1.0)

        Returns:
            DataFrame with loaded data
        """
        filepath = self.data_path / filename

        if not filepath.exists():
            raise FileNotFoundError(f"Data file not found: {filepath}")

        if sample_frac is not None and not 0.0 < sample_frac <= 1.0:
            raise ValueError(f"sample_frac must be in (0, 1], got {sample_frac}")

        df = pd.read_csv(filepath)

        if sample_frac is not None and sample_frac < 1.0:
            df = df.sample(frac=sample_frac, random_state=42)

        return df
```

## Code Snippet: BaseDetector

```python
# src/detection/base_detector.py
"""Base class for anomaly detectors."""

from abc import ABC, abstractmethod
import numpy as np


class BaseDetector(ABC):
    """Abstract base class for all anomaly detectors."""

    def __init__(self, config: dict):
        self.config = config
        self.is_fitted = False

    @abstractmethod
    def fit(self, X: np.ndarray) -> 'BaseDetector':
        """Fit detector to training data."""
        pass

    @abstractmethod
    def score_samples(self, X: np.ndarray) -> np.ndarray:
        """Compute anomaly scores (higher = more anomalous)."""
        pass

    def predict(self, X: np.ndarray, threshold: float) -> np.ndarray:
        """Predict anomaly labels (1 = anomaly, 0 = normal)."""
        return (self.score_samples(X) > threshold).astype(int)
```

## Verification

```bash
# After implementation, verify with:
python -c "
from src.ingestion.data_loader import DataLoader
from src.detection.isolation_forest import IsolationForestDetector
import numpy as np

# Test data loader
loader = DataLoader({'data': {'path': 'data/'}})
print('DataLoader: OK')

# Test Isolation Forest
config = {'detection': {'isolation_forest': {'n_estimators': 50}}}
detector = IsolationForestDetector(config)
X = np.random.randn(100, 5)
detector.fit(X)
scores = detector.score_samples(X)
print(f'IsolationForest: OK (scores shape: {scores.shape})')
"

# Run tests
pytest tests/test_data_loader.py -v
```

## Definition of Done

- [ ] All files created and importable
- [ ] Data loads without memory error (10% KDD sample)
- [ ] Preprocessing handles missing values correctly
- [ ] Isolation Forest returns scores where anomalies > inliers
- [ ] Unit tests pass (`pytest tests/test_data_loader.py`)
- [ ] Configuration loads from YAML

## Learning Notes

**Concepts to understand:**
- Abstract Base Classes (ABC) in Python
- sklearn `fit`/`transform`/`predict` pattern
- YAML configuration with PyYAML
- pytest fixtures and assertions
